### ISSUE TITLE
Armorstand equipment lock/unlock feature

### ISSUE THREAD

**Issue by Sikatsu** [Wolfst0rm/ArmorStandEditor-Issues#5]
### Is your feature request related to a problem?

Is it possible to add the lock/unlock equipment feature from this fork?

https://github.com/JessePesse123/ArmorStandEditor/releases/tag/v1.17-0

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

**Comment by Sikatsu** [comment]
I got a suggestion from a player for this feature.

In case the armorstand's equipment is locked, that if you target the armorstand with F (offhand) the glow will be gray rather then white. To indicate the equipment of this armorstand is indeed locked.

**Comment by Wolfieheart** [comment]
Hey thanks for requesting this. Has been accepted and work will start ASAP

**Comment by Wolfieheart** [comment]
Hrmmm looking into this it appears that it was already implemented, just not documented accordingly. Working on Fixing all now :)

**Comment by Sikatsu** [comment]
Oh actually? It wasn't an option in the flint menu haha

**Comment by Wolfieheart** [comment]
Yeah just checked the class that it involves - https://github.com/Wolfst0rm/ArmorStandEditor/blob/master/src/main/java/io/github/rypofalem/armorstandeditor/menu/Menu.java#L141

and

https://github.com/Wolfst0rm/ArmorStandEditor/blob/master/src/main/java/io/github/rypofalem/armorstandeditor/PlayerEditor.java#L310

Apparently it was implemented but not documented. Will get a fix soon and check that it works properly. Noticed as well some things I didn't fix when doing the switch for ItemFrames to asedit.invisible as well :D

**Comment by Wolfieheart** [comment]
I dont think the targeting color is implemented however. So I will investigate that a bit further as well :D

**Comment by Wolfieheart** [comment]
Hrmmm the targetting Color is not directly implementable with PotionEffect. 
Could have to do something with Scoreboards in that case, Looking into it further.

**Comment by Sikatsu** [comment]
I believe it has to do with scoreboards indeed, if I am right clicking F currently makes the armorstand glow white? So maybe look into the method that was used for that? ^^

**Comment by Wolfieheart** [comment]
The method used for that is for when we attempt to Target, we (at the end of targeting) highlight it as so:

```
private void highlight(ArmorStand armorStand) {
		armorStand.removePotionEffect(PotionEffectType.GLOWING);
		armorStand.addPotionEffect(new PotionEffect(PotionEffectType.GLOWING, 15, 1, false, false));
	}
```

Issue is with PotionEffect there is no way to color it as far as I know from the class PotionEffect itself. So investigating the Scoreboard option.

**Comment by Wolfieheart** [comment]
The first part on DisableSlots is in if you want to go ahead and test that: https://github.com/Wolfst0rm/ArmorStandEditor/suites/3628315516/artifacts/87788238

Let me know here if something doesn't work correctly :D 

I will later on investigate the issue RE: Scoreboard and adding that for Glowing Effects :D

**Comment by Wolfieheart** [comment]
Alrighty working on the implementation of colors and scoreboards now :D

**Comment by Sikatsu** [comment]
Awesome! I'll test it out later today since i'm out in the city at the moment. Thanks for your work on this, we appreciate it!

**Comment by Wolfieheart** [comment]
> Awesome! I'll test it out later today since i'm out in the city at the moment. Thanks for your work on this, we appreciate it!

No worries. Thank you for reporting it and for the comments as well :D 

Have a good time in the city, and I will continue working on it when I have a free moment (also at work at the moment)

**Comment by Sikatsu** [comment]
Just reached home, about to test this out! ^^

**Comment by Wolfieheart** [comment]
Discovered an issue when loading the plugin. The latest build I am testing at the moment to ensure the scoreboards work etc.

**Comment by Wolfieheart** [comment]
Alrighty - Tested with the latest artifact. Scoreboard too many characters issue should be fixed. Feel free to give the latest build artifact a shot :) 

You can find it here: https://github.com/Wolfst0rm/ArmorStandEditor/suites/3630531403/artifacts/87864927

**Comment by Sikatsu** [comment]
The other day I saw an armorstand editor on spigot that allows you to pickup and place armorstands a anytime. Even when they are tweaked, do you think this would be a nice addition? It's not like the save/copy system but actually picking it up and that it remembers everything the player changed.

**Comment by Sikatsu** [comment]
Will test it out! :)

**Comment by Wolfieheart** [comment]
> The other day I saw an armorstand editor on spigot that allows you to pickup and place armorstands a anytime. Even when they are tweaked, do you think this would be a nice addition? It's not like the save/copy system but actually picking it up and that it remembers everything the player changed.

Might be something worth investigating at some point.

> Will test it out! :)

Feel free to use this for any issues then I cant test locally atm, so I will have to test it later this week when I have free time :)

**Comment by Sikatsu** [comment]
I think the colors are reversed, so unlocked = gray and locked = white.
Maybe an even better or atleast more clear color could be red (or black)?

**Comment by Wolfieheart** [comment]
> I think the colors are reversed, so unlocked = gray and locked = white.
> Maybe an even better or atleast more clear color could be red (or black)?

Hrmm ok. That might be my mistake as to where to add and remove teams.

We could make it so that when its on the scoreboard ASLocked that its Red for Locked, then create a new scoreboard for Unlocked that is Black. 

Then question is when we untarget, what do we want the color to be.

**Comment by Sikatsu** [comment]
It also seems ALL actions show gray glow on the armorstand at the moment, rather then just the disabled slots action making the armorstand turn gray when locked and white when unlocked.

If you'd like to have glow for all actions, I think an idea would be to use green for success. White for targetting an armorstand with F, red for locked equipment/slots?

**Comment by Sikatsu** [comment]
I don't know what happened but the glow stopped working just now except for when I target the armorstand with F

**Comment by Wolfieheart** [comment]
> It also seems ALL actions show gray glow on the armorstand at the moment, rather then just the disabled slots action making the armorstand turn gray when locked and white when unlocked.

> I don't know what happened but the glow stopped working just now except for when I target the armorstand with F

These might require further debugging. Will have to investigate soon when I can get access to my debug env :)

> If you'd like to have glow for all actions, I think an idea would be to use green for success. White for targetting an armorstand with F, red for locked equipment/slots?

Was just loud thinking on my part. We can change it to be more visible.
Ideally, Locked is Red and Unlocked is White is a good idea IMO.

Would require 1 extra scoreboard assignment in OnEnable and OnDisable for Registering and Unregistering, therefore also assignments for removing it from one team and adding it to another. Shouldnt be too hard to accomplish.

**Comment by Wolfieheart** [comment]
Also, I would probably go this for colors:

Red for Lock
White as it is now for Targetting and Unlocking - Black for Targetting/Unlocking in a Dark Area won't be seen.

**Comment by Wolfieheart** [comment]
Feel free to review the latest commit with regards to the Colors btw :) Will help me discover what color goes with what functionality of the plugin, as I am still not sure how this part works exactly, as in what is not target/targeted:

https://github.com/Wolfst0rm/ArmorStandEditor/blob/master/src/main/java/io/github/rypofalem/armorstandeditor/PlayerEditor.java#L411

**Comment by Sikatsu** [comment]
Awesome, will check around now! Give me a moment haha ^^

**Comment by Sikatsu** [comment]
I was afk for a bit and after I came back the glow feature works again.

Most actions like arms, visibility, size, plate are directly visible so I don't think they require glow.

But the equipment slot action could use glow as a confirmation whether something is on/off for that armorstand.

I think the colors sound right, if you happen to have a jar I could test let me know :)

**Comment by Wolfieheart** [comment]
> But the equipment slot action and the gravity feature could use glow as a confirmation whether something is on/off for that armorstand.
> I think the colors sound right, if you happen to have a jar I could test let me know :)

JAR should be here: https://github.com/Wolfst0rm/ArmorStandEditor/runs/3462035501

> But the equipment slot action and the gravity feature could use glow as a confirmation whether something is on/off for that armorstand.

Then Proposal that Red is ON and White is OFF going forward, for sake of ease since scoreboards are already implemented?

**Comment by Sikatsu** [comment]
I seem to receive this error: https://pastebin.com/u1kSDpj4

And yeah, that sounds about right! Currently don't seem to have glow at all though due to the error above o.O

**Comment by Wolfieheart** [comment]
> I seem to receive this error: https://pastebin.com/u1kSDpj4

Interesting since that deals with getting the UUID of the ArmorStandEntity. So interesting that it cant get the Unique ID for that entity, and since the team is null it cant do anything. Hrmmm. 

Was that happening with the build you used earlier or just now started?

**Comment by Sikatsu** [comment]
The most recent build. After a while the glow starts working though.
As for the disabledslots setting, locked = red and unlocked = white. That seems correct!
Maybe this can also be applied to the gravity setting.

Also, all other actions always give white glow as well when you use them on an armorstand. If you know what I mean?

I can always send a video on discord if needed haha

**Comment by Wolfieheart** [comment]
Ok then I'll have to deep dive. A video is nice to have as well since It allows me to see what is going on and where therefore when debugging I can break in and see whats what. 

If you want to send the link to it here if you upload it to Discord, that works fine. Will move this from Implemented to To Be Tested P1

**Comment by Sikatsu** [comment]
Does this work?
https://cdn.discordapp.com/attachments/184716181263876096/881900756419084318/Minecraft__1.17.1_-_Multiplayer_3rd-party_Server_2021-08-30_15-56-52.mp4

If yes, as you can see the glow works on every action I do depending on what the locked/unlocked setting is set too ^^

**Comment by Wolfieheart** [comment]
Alrighty I am now home and I am debugging using Paper 1.17.1 Build 196 (so a bit outdated but still fine since latest).
The video works, thank you for that! Its a great help.

So the disabled slots colors are correct :D HUZZAH!

Testing with the ArmorStand Targetting now : D

**Comment by Sikatsu** [comment]
Welcome home! They are indeed correct, but the glow seems to show up with every action ^^

**Comment by Wolfieheart** [comment]
Hrmmm I am willing to leave it out for Targetting until we can look at it later :) 

I am glad however that Disabling of Slots is working in the meantime :D

**Comment by Sikatsu** [comment]
I am not sure if I understood you, I don't think it should show the white glow when you use an action on an armorstand? Only when you target the armorstand with the offhand key F?

Any luck in checking my other issues? ^^

**Comment by Wolfieheart** [comment]
Well from what I understand it should be as follows:

Red for Lock
White for Unlocking - Nothing else should be affected by this.

Investigating it more now - Just finished dinner

**Comment by Wolfieheart** [comment]
Alrighty testing the Disable Colors. I realised a few of the issues. ALSO, I know how the targeting works more now. and I know which part of the ArmorStand does Disable and Enable. Adding Comments and working on the Colors again to be 100% sure :D

**Comment by Sikatsu** [comment]
Awesome! Thanks so much.

**Comment by Wolfieheart** [comment]
Ok so going to explain it now that I have tested it (and changed a few things):

So it will do the following:

- When it is locked and your press F on the ArmorStand - Glow Red for 15 seconds 
- When it is Unlocked and you press F on the ArmorStand - Glow White for 15 seconds
- When Locked by Right Click to Damage - Unlock and Glow White for 60 seconds
- When unlocked and Right Clicked - Lock and Glow Red for 60 seconds

That seems OK with me. Going to commit this and let you test it again to be sure :D

**Comment by Wolfieheart** [comment]
If you want to test it you can via: https://github.com/Wolfst0rm/ArmorStandEditor/suites/3633931694/artifacts/87989325

But for me this is good 👍 to publish to production

**Comment by Sikatsu** [comment]
Testing right away, thanks for the explanation also!

**Comment by Sikatsu** [comment]
It appears to be one second for me, rather than 15 seconds.
What exactly did you mean with right click to damage?
And when I lock/unlock it, it seems to be just 1 second as well.

**Comment by Sikatsu** [comment]
Also after I locked/unlocked an armorstand, all other actions I perform on an armorstand show red or white glow too depending on the lock/unlocked state, but I think this is not needed?

**Comment by Wolfieheart** [comment]
> What exactly did you mean with right click to damage?

Thats my bad - I meant when the ArmorStand is Damaged at all (so left click). I am tired lol.

> And when I lock/unlock it, it seems to be just 1 second as well.

Ah I misunderstood the duration part of Potion Effect since its in Ticks. Will look at it.

**Comment by Wolfieheart** [comment]
> Also after I locked/unlocked an armorstand, all other actions I perform on an armorstand show red or white glow too depending on the lock/unlocked state, but I think this is not needed?

Will investigate this shortly.

**Comment by Wolfieheart** [comment]
Initial Investigation: Could it potentially be related to this?

https://github.com/Wolfst0rm/ArmorStandEditor/blob/4331a386b98f81f3c6cd15437d226a4423e52f75/src/main/java/io/github/rypofalem/armorstandeditor/PlayerEditor.java#L500

**Comment by Sikatsu** [comment]
Possibly, it seems to highlight the armorstand but only after you locked or unlocked the equipment slots :o

**Comment by Wolfieheart** [comment]
Yep its linked to that. Commented out and committing - Thanks for pointing that out to me :D Committing the fix if you want to do one final test :D

**Comment by Wolfieheart** [comment]
Commited and ready to be checked :D https://github.com/Wolfst0rm/ArmorStandEditor/suites/3634261221/artifacts/88000702

**Comment by Sikatsu** [comment]
It seems to work now! Altho 60 seconds feels very long, i'd personally make it 15 or less as well. Good job on this! :D

**Comment by Wolfieheart** [comment]
Yeah you are right. I will go ahead and make that change and commit it and close this :)

**Comment by Sikatsu** [comment]
Awesome! I have a few more small suggestions, do you want me to make new issues for those?

**Comment by Wolfieheart** [comment]
Actually if they are related to this go ahead and do them here. Otherwise feel free to open a new issue for those :D

**Comment by Wolfieheart** [comment]
Marking this as Tested and Fixed for sake of clarity.

Pending Implementation for the suggestions :)

**Comment by Sikatsu** [comment]
Hmm no they are not, it's related to the translation system. So I can use color codes / hexcodes for specific words in a line, the current system feels very old where you can only have one color code per line. I also can't seem to translate the words on/off :o

**Comment by Wolfieheart** [comment]
Feel free to submit another Issue for that and I will take a look into that. Will probably need some help for that eventually :) However since they are unrelated, I will close this. Thanks for reporting this!

**Comment by Sikatsu** [comment]
I noticed if I change the armorstand target with F, the old armorstand also still gets highlighted for the remaining time.
![image](https://user-images.githubusercontent.com/5664135/131397271-5232246e-0989-4172-9f3a-612438280357.png)


Would it be an idea if the target was changed that the old armorstand stops highlighting immediately?
Or, the time of highlighting with F could be made much lower (similiar to before)

**Comment by Wolfieheart** [comment]
Yeah I might make the highlighting timing lower from say 15 seconds. Its easier that way I think.

Ideas on what we could time it as globally? I would do 60ticks or 3 seconds so its not permanent

**Comment by Sikatsu** [comment]
I do think 2-3 seconds is enough to see what's happening ^^

**Comment by Wolfieheart** [comment]
Then i will go ahead and implement that and keep the issue closed :)
